sklearn/decomposition/pca.py

```diff
@@ -251,50 +320,68 @@ def fit_transform(self, X, y=None):
+            if n_components >= 1 and n_components < .8 * min(X.shape):
+                svd_solver = 'randomized'
+            else:
+                svd_solver = 'full'
+
+        # Call different fits for either full or truncated SVD
+        if svd_solver == 'full':
+            return self._fit_full(X, n_components)
+        elif svd_solver in ['arpack', 'randomized']:
+            return self._fit_truncated(X, n_components, svd_solver)
+
+    def _fit_full(self, X, n_components):
+        """Fit the model by computing full SVD on X"""
         n_samples, n_features = X.shape
-        X = as_float_array(X, copy=self.copy)
-        # Center data
-        self.mean_ = np.mean(X, axis=0)
-        X -= self.mean_
-        U, S, V = linalg.svd(X, full_matrices=False)
-        explained_variance_ = (S ** 2) / n_samples
-        explained_variance_ratio_ = (explained_variance_ /
-                                     explained_variance_.sum())
 
-        components_ = V
-
-        n_components = self.n_components
-        if n_components is None:
-            n_components = n_features
-        elif n_components == 'mle':
+        if n_components == 'mle':
             if n_samples < n_features:
                 raise ValueError("n_components='mle' is only supported "
                                  "if n_samples >= n_features")
-
-            n_components = _infer_dimension_(explained_variance_,
-                                             n_samples, n_features)
         elif not 0 <= n_components <= n_features:
-            raise ValueError("n_components=%r invalid for n_features=%d"
+            raise ValueError("n_components=%r must be between 0 and "
+                             "n_features=%r with svd_solver='full'"
                              % (n_components, n_features))
 
-        if 0 < n_components < 1.0:
-            # number of components for which the cumulated explained variance
-            # percentage is superior to the desired threshold
+        # Center data
+        self.mean_ = np.mean(X, axis=0)
+        X -= self.mean_
+
+        U, S, V = linalg.svd(X, full_matrices=False)
+        # flip eigenvectors' sign to enforce deterministic output
+        U, V = svd_flip(U, V)
+
+        components_ = V
+
+        # Get variance explained by singular values
+        explained_variance_ = (S ** 2) / n_samples
+        total_var = explained_variance_.sum()
+        explained_variance_ratio_ = explained_variance_ / total_var
+
+        # Postprocess the number of components required
+        if n_components == 'mle':
+            n_components = \
+                _infer_dimension_(explained_variance_, n_samples, n_features)
+        elif 0 < n_components < 1.0:
+            # number of components for which the cumulated explained
+            # variance percentage is superior to the desired threshold
             ratio_cumsum = explained_variance_ratio_.cumsum()
             n_components = np.sum(ratio_cumsum < n_components) + 1
```

Comment:
It seems to always be faster, and scales as O[log N] rather than O[N]:
```python
for N in 10 ** np.arange(1, 8, 2):
    x = np.cumsum(np.random.rand(N))
    x /= x[-1]
    print("\nN =", N)
    %time (x < 0.9).sum()
    %time np.searchsorted(x, 0.9)
```

```
N = 10
CPU times: user 32 µs, sys: 0 ns, total: 32 µs
Wall time: 35 µs
CPU times: user 10 µs, sys: 4 µs, total: 14 µs
Wall time: 16 µs

N = 1000
CPU times: user 25 µs, sys: 11 µs, total: 36 µs
Wall time: 29.8 µs
CPU times: user 9 µs, sys: 3 µs, total: 12 µs
Wall time: 12.9 µs

N = 100000
CPU times: user 351 µs, sys: 125 µs, total: 476 µs
Wall time: 363 µs
CPU times: user 10 µs, sys: 1 µs, total: 11 µs
Wall time: 12.2 µs

N = 10000000
CPU times: user 18.8 ms, sys: 1.95 ms, total: 20.7 ms
Wall time: 20.8 ms
CPU times: user 19 µs, sys: 1 µs, total: 20 µs
Wall time: 23.1 µs
```